### PR TITLE
Refactor TestProcessorNode

### PR DESCRIPTION
**Pull request by sstanculeanu**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- a lot of duplicated code was used for integration tests
  
## Proposed Changes
- add `ArgTestProcessorNode` and create the node for tests based on provided values
- remove as much as possible the duplicated constructors

## Testing procedure
- Integration tests code only

### CODE REVIEW

File: integrationTests/testInitializer.go [hunk 1]
```diff
@@ -1497,7 +1495,14 @@ func CreateNodesWithCustomStateCheckpointModulus(
 	idx := 0
 	for shardId := uint32(0); shardId < uint32(numOfShards); shardId++ {
 		for j := 0; j < nodesPerShard; j++ {
-			n := NewTestProcessorNodeWithStateCheckpointModulus(uint32(numOfShards), shardId, shardId, stateCheckpointModulus)
+			n := NewTestProcessorNode(ArgTestProcessorNode{
+				MaxShards:            uint32(numOfShards),
+				NodeShardId:          shardId,
+				TxSignPrivKeyShardId: shardId,
+				StateCheckpointModulus: &StateCheckpointModulus{
```
Comment (iulianpascalau) [inline integrationTests/testInitializer.go:1502]: interesting approach to be able to check "unset" primitive values.
Maybe we should have renamed the struct `StateCheckpointModulus` to `IntWrapper` so we could have used it elsewhere? We can have a .go file for all used wrappers like this one. 
Also, please check the indentation (L1501 vs L1502)

File: integrationTests/testProcessorNode.go [hunk 2]
```diff
@@ -761,41 +585,99 @@ func (tpn *TestProcessorNode) initValidatorStatistics() {
...
+	tpn.initStorage()
+	if check.IfNil(args.TrieStore) {
+		tpn.initAccountDBsWithPruningStorer()
+	} else {
+		tpn.initAccountDBs(args.TrieStore)
+	}
+
+	economicsConfig := tpn.createDefaultEconomicsConfig()
+	if isFullGenesis {
+		economicsConfig.GlobalSettings.YearSettings = append(
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNode.go:606]: also here: we set some economics config based on condition that is unrelated

File: integrationTests/testProcessorNode.go [hunk 3]
```diff
@@ -761,41 +585,99 @@ func (tpn *TestProcessorNode) initValidatorStatistics() {
...
+			tpn.ShardCoordinator,
+			tpn.Storage,
+			tpn.BlockChain,
+			tpn.DataPool,
+			tpn.EconomicsData,
+			&genesisMocks.AccountsParserStub{},
+			tpn.SmartContractParser,
+			tpn.EnableEpochs,
+		)
+	} else if args.WithSync {
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNode.go:640]: maybe we can refactor the else-if branches? 
extract in a function, perhaps?

File: integrationTests/testProcessorNodeWithMultisigner.go [hunk 4]
```diff
@@ -561,20 +445,35 @@ func CreateNodesWithNodesCoordinatorAndHeaderSigVerifier(
...
-				cp,
-				i,
-				nil,
-				headerSig,
-				createHeaderIntegrityVerifier(),
-				nodesSetup,
-			)
+			multiSigner, err := createMultiSigner(*cp, shardId, i)
+			if err != nil {
+				fmt.Println("Error generating multisigner")
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNodeWithMultisigner.go:456]: log.Error?

**Review by iulianpascalau — COMMENTED** [review]
Very good refactor! :+1:

File: integrationTests/testProcessorNodeWithCoordinator.go [hunk 5]
```diff
@@ -101,17 +99,51 @@ func CreateProcessorNodesWithNodesCoordinator(
...
+
+			pubKeysMap := pubKeysMapFromKeysMap(ncp)
+			kp := ncp[shardId][i]
+			multiSigner, err := multisig.NewBLSMultisig(
+				llsig,
+				pubKeysMap[shardId],
+				kp.BlockSignSk,
+				kp.BlockSignKeyGen,
+				uint16(i),
 			)
```
Comment (ssd04) [inline integrationTests/testProcessorNodeWithCoordinator.go:113]: can we use the `createMultiSigner` function from the other file (withMultiSigner) here also?

File: integrationTests/testProcessorNode.go [hunk 6]
```diff
@@ -416,260 +409,88 @@ func newBaseTestProcessorNode(
...
-	tpn.OwnAccount = CreateTestWalletAccount(shardCoordinator, txSignPrivKeyShardId)
-
-	tpn.initDataPools()
-	if tpn.ShardCoordinator.SelfId() != core.MetachainShardId {
-		tpn.DataPool = dPool
-	}
-
-	tpn.initTestNode()
+// NewTestProcessorNode returns a new TestProcessorNode instance with a libp2p messenger
+func NewTestProcessorNode(args ArgTestProcessorNode) *TestProcessorNode {
```
Comment (ssd04) [inline integrationTests/testProcessorNode.go:491]: is the comment still relevant? regarding "with a libp2p messenger", now it's more generic, right?

File: integrationTests/singleShard/block/interceptedRequestHdr/interceptedRequestHdr_test.go [hunk 7]
```diff
@@ -34,10 +34,18 @@ func TestNode_GenerateSendInterceptHeaderByNonceWithNetMessenger(t *testing.T) {
 	var txSignPrivKeyShardId uint32 = 0
 
 	fmt.Println("Requester:	")
-	nRequester := integrationTests.NewTestProcessorNode(nrOfShards, shardID, txSignPrivKeyShardId)
+	nRequester := integrationTests.NewTestProcessorNode(integrationTests.ArgTestProcessorNode{
+		MaxShards:            nrOfShards,
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply integrationTests/singleShard/block/interceptedRequestHdr/interceptedRequestHdr_test.go]: done

File: integrationTests/testInitializer.go (hunk 1 above)
Comment (sstanculeanu) [reply integrationTests/testInitializer.go]: added `IntWrapper` as suggested

File: integrationTests/testProcessorNode.go [hunk 8]
```diff
@@ -416,260 +409,88 @@ func newBaseTestProcessorNode(
...
+			OptimizeGasUsedInCrossMiniBlocksEnableEpoch: UnreachableEpoch,
+			ScheduledMiniBlocksEnableEpoch:              UnreachableEpoch,
+			MiniBlockPartialExecutionEnableEpoch:        UnreachableEpoch,
+			FailExecutionOnEveryAPIErrorEnableEpoch:     UnreachableEpoch,
+		}
+
+		isFullGenesis := args.GenesisFile != ""
+		isWithStateCheckpointModulus := args.StateCheckpointModulus != nil
+		if isFullGenesis || isWithStateCheckpointModulus {
+			epochsConfig.StakingV2EnableEpoch = UnreachableEpoch
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply integrationTests/testProcessorNode.go]: as suggested, now tests with full genesis or with state checkpoint modulus pass a custom enable epochs config

File: integrationTests/testProcessorNode.go (hunk 2 above)
Comment (sstanculeanu) [reply integrationTests/testProcessorNode.go]: done, now economics config is part of `ArgTestProcessorNode`

File: integrationTests/testProcessorNodeWithCoordinator.go (hunk 5 above)
Comment (sstanculeanu) [reply integrationTests/testProcessorNodeWithCoordinator.go]: done

File: integrationTests/testProcessorNode.go (hunk 6 above)
Comment (sstanculeanu) [reply integrationTests/testProcessorNode.go]: the part `with a libp2p messenger` is still valid, but I updated the comment

**Review by ssd04 — APPROVED** [review]

File: integrationTests/testProcessorNodeWithCoordinator.go [hunk 9]
```diff
@@ -99,23 +81,14 @@ func CreateProcessorNodesWithNodesCoordinator(
...
-				llsig,
-				pubKeysMap[shardId],
-				kp.BlockSignSk,
-				kp.BlockSignKeyGen,
-				uint16(i),
-			)
+			multiSigner, err := createMultiSigner(*cp, shardId, i)
 			if err != nil {
-				fmt.Printf("error generating multisigner: %s\n", err)
+				log.Error("error generating multisigner: %s\n", err)
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNodeWithCoordinator.go:86]: no need for the extra "\n". Valid here and on the other 3 locations
Could have been: 
```GO
log.Error("error generating multisigner", "error", err)
```

**Review by iulianpascalau — APPROVED** [review]
